Bloom.md

```diff
@@ -0,0 +1,248 @@
+* Performance
+* Popularity
+
+https://crates.io/crates/bloomfilter was chosen as it provides all the required APIs for controlling a bloom filter. This library has also been stable through the releases.
+
+Certain libraries (e.g. fastbloom) are no longer compatible with older versions (after updates) resuling in serialization / deserialization incompatibility.
+
+growable-bloom-filter is a bloom filter library that implements a scalable bloom filter. However, it does not allow control over when the scaling occurs and does not expose APIs to check the number of filters that it has scaled out to. This crate also has had a breaking change across versions and bloom filters of the older version are not loadable on the newer versions.
+
+
+### Large BloomFilter objects
+
+Create and Delete operations on large bloom filters take longer durations and will block the main thread for this duration. 
+Because of this, the following operations can be handled differently:
+* defrag callback: If the memory used is greater than X MB (configurable with `bf.bloom_large_item_threshold`), we can skip defrag operations on this bloom object.
+* free_effort callback: This callback decides the free effort for the bloom object. If it is greater than X MB (configurable with `bf.bloom_large_item_threshold`), we can return 0 to use async free logic.
+* create operations (BF.ADD/MADD/INSERT): We can compute the estimated size of the bloom filter that will be created as a result of the operation. We can consider rejecting requests in case of unavailable memory.
+
+
+## Specification
+
+### Bloom Filter Command API
+
+The following Bloom Filter commands are supported and their API syntax is compatible with ReBloom:
+BF.ADD <key> <item>
+BF.EXISTS <key> <item>
+BF.MADD <key> <item> [<item> ...]
+BF.MEXISTS <key> <item> [<item> ...]
+BF.CARD <key>
+BF.INFO <key> [CAPACITY | SIZE | FILTERS | ITEMS | EXPANSION]
+BF.RESERVE <key> <false_positive_rate> <capacity> [EXPANSION <expansion>] | [NONSCALING]
+BF.INSERT <key> [CAPACITY <capacity>] [ERROR <fp_error>] [EXPANSION <expansion>] [NOCREATE] [NONSCALING] ITEMS <item> [<item> ...]
+
+Currently following commands (from ReBloom) are not supported:
+BF.LOADCHUNK <key> <iterator> <data>
+BF.SCANDUMP <key> <iterator>
+
+The BF.SCANDUMP command is used to perform an incremental save on specific Bloom filter object.
+The BF.LOADCHUNK is used to incrementally load / restore a bloom filter object from data from the BF.SCANDUMP command.
+
+The reason for not implementing these two commands is because the Module provides the ability to load and save BloomModule data type items during RDB load and save. BF.LOADCHUNK and BF.SCANDUMP are APIs to load bloom filter objects through commands, but since we will provide RDB save & load, having specific commands for the same purpose was not considered as required.
+
+Note: Based on the AOF Rewrite discussion, we will decide on whether these commands should be supported.
```

Comment:
Does scandump work for AOF? We need to dump the entire object as a single object int he AOF file. Could we instead just have a BF.LOAD command? I suppose we should figure out what RedisBloom does.

Reply:
a related question: what is the maximum a bloom object can use? Are there any limits in the implementation or design?

Reply:
>Could we instead just have a BF.LOAD command?

@madolson The "AOF Rewrite handling" has been updated to explain the usage of BF.LOAD during AOF Rewrite.

>a related question: what is the maximum a bloom object can use? Are there any limits in the implementation or design?

@parthpatel This is explained more in the "Large BloomFilter objects" section